### PR TITLE
Start testing with github actions

### LINKED ISSUE #198: Travis -> GH Actions
It would be nice to migrate from Travis to GH actions.

### PR DESCRIPTION
**Pull request by TysonAndre**
The script `ci/test_dockerized.sh 8.0` can be used to run the exact same
tests locally.

Closes #198

### CODE REVIEW

**Comment by TysonAndre** [comment]
Sorry for the noise, still getting familiar with github actions, it's working now that I pushed a branch to this upstream rather than my fork

https://github.community/t/github-action-not-triggered-on-pull-request/18009

> Is your upstream repo a private one? There is an limitation for forked repo. Workflows do not run on private base repositories when you open a pull request from a forked repository.
> 
> https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories

File: ci/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ARG PHP_VERSION
+FROM php:$PHP_VERSION
+WORKDIR /php-ast
+ADD *.c *.h *.php config.m4 config.w32 package.xml LICENSE README.md ./
+
+# Assume compilation will be the time consuming step.
+# Add tests after compiling so that it's faster to update tests and re-run them locally.
+RUN phpize && CFLAGS='-Wall -O2 -Wextra' ./configure && make -j2 && make install
```
Comment (nikic) [inline ci/Dockerfile]: Why it is necessary to explicitly pass `-O2`?

File: ci/test_inner.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+# -x Exit immediately if any command fails
+# -e Echo all commands being executed.
+# -u fail for undefined variables
+set -xeu
+echo "Run tests in docker"
+php ./run-tests.php -P -q --show-diff
```
Comment (nikic) [inline ci/test_inner.sh:7]: Does this need REPORT_EXIT_STATUS as well?

File: package.xml [hunk 3]
```diff
@@ -30,7 +30,6 @@
  <license uri="https://github.com/nikic/php-ast/blob/master/LICENSE">BSD-3-Clause</license>
  <notes>
 - Support php 8.1 enums, add 'type' (for enum type) to AST_CLASS nodes in AST version 85+.
-- Support php 8.1 'never' return type.
```
Comment (nikic) [inline package.xml:33]: Accident?

**Review by nikic — COMMENTED** [review]
Meta question: Why do we need to go through docker?

File: ci/Dockerfile (hunk 1 above)
Comment (TysonAndre) [reply ci/Dockerfile]: The optimization choice was arbitrary - I just chose something that would be similar to what end users would use

File: ci/test_inner.sh (hunk 2 above)
Comment (TysonAndre) [reply ci/test_inner.sh:7]: Oh. I added that to the wrong command

File: package.xml (hunk 3 above)
Comment (TysonAndre) [reply package.xml:33]: I accidentally saved the old version in my editor (started before switching)

File: ci/Dockerfile (hunk 1 above)
Comment (nikic) [reply ci/Dockerfile]: What I meant was, isn't this what you get anyway because you're building against a release PHP build?

Comment (TysonAndre) [reply ci/Dockerfile]: It doesn't seem like it - `CFLAGS=... ./configure` seems like it's overriding the cflags, not adding to them when I look at the generated Makefile.

And I wanted to add the `-Wall -Wextra` to be able to see if a given change introduces compiler warnings (in the web ui), despite it not affecting the test results or compilation success

**Comment by TysonAndre** [comment]
> Meta question: Why do we need to go through docker?

It takes 25-30 seconds in Travis and allows contributors/maintainers to see if a change would fix the GH actions build before pushing it.

It's convenient. Unlike travis, github actions does not provide prebuilt php images (that I know of), it just has images for various OSes with a large number of preinstalled utilities.
So if I want to install php 7.0 - I may be out of luck entirely or need to install a PPA, which would probably end up being even slower than the docker step (and one would probably dockerize that step to run it locally)

- Installing php 7.0 would also be brittle even if I got it working - e.g. if I needed to test in php 7.0, but I could only get it to work in ubuntu 16.04 due to library changes, and GH actions ended up dropping support for it in 16.04. (I haven't actually tried it, but I'd expect issues like that to come up now or in the future)

File: ci/Dockerfile (hunk 1 above)
Comment (TysonAndre) [reply ci/Dockerfile]: Switched to EXTRA_CFLAGS - this seems to exist in the generated Makefile, even in PHP 7.0

File: ci/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,16 @@
+ARG PHP_VERSION
+FROM php:$PHP_VERSION
+WORKDIR /php-ast
+ADD *.c *.h *.php config.m4 config.w32 package.xml LICENSE README.md ./
+
+# Assume compilation will be the time consuming step.
+# Add tests after compiling so that it's faster to update tests and re-run them locally.
+#
+# The choice of optimization flags is arbitrary and shouldn't affect test results.
+# Higher optimization levels will slow compilation but speed up tests (especially if valgrind (run-tests.php -m) is ever used)
```
Comment (nikic) [inline ci/Dockerfile:10]: These comments are outdated now that EXTRA_CFLAGS is used.

**Review by nikic — APPROVED** [review]

File: ci/Dockerfile (hunk 4 above)
Comment (TysonAndre) [reply ci/Dockerfile]: I removed them

**Pull request merged**